### PR TITLE
Draft changes for issue 4145

### PR DESCRIPTION
**Pull request by jlucovsky**
This commit is *draft* and demonstrates an approach to common registrations for file data.

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [4145](https://redmine.openinfosecfoundation.org/issues/4145)

Describe changes:
- Build on 4141 use common handling for all file based detection

With changes:
```
$ hyperfine -m3 'sudo rm -rf /tmp/ll/* && ./src/suricata -c suricata.yaml -l /tmp/ll -S ~/suricata.rules -r ~/TLP_WHITE-smb-16kfiles-no-close.pcap --set outputs.1.eve-log.threaded=true'
Benchmark 1: sudo rm -rf /tmp/ll/* && ./src/suricata -c suricata.yaml -l /tmp/ll -S ~/suricata.rules -r ~/TLP_WHITE-smb-16kfiles-no-close.pcap --set outputs.1.eve-log.threaded=true
  Time (mean ± σ):     1506.882 s ± 30.635 s    [User: 1522.916 s, System: 65.111 s]
  Range (min … max):   1476.483 s … 1537.747 s    3 runs
  ```

Comparative numbers from `master`:
```
$ hyperfine -m3 'sudo rm -rf /tmp/ll/* && ./src/suricata -c suricata.yaml -l /tmp/ll -S ~/suricata.rules -r ~/TLP_WHITE-smb-16kfiles-no-close.pcap --set outputs.1.eve-log.threaded=true'
Benchmark 1: sudo rm -rf /tmp/ll/* && ./src/suricata -c suricata.yaml -l /tmp/ll -S ~/suricata.rules -r ~/TLP_WHITE-smb-16kfiles-no-close.pcap --set outputs.1.eve-log.threaded=true
  Time (mean ± σ):     1555.037 s ± 25.689 s    [User: 1577.058 s, System: 62.196 s]
  Range (min … max):   1528.016 s … 1579.144 s    3 runs
  ```
### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the pull request number.

```
SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=pr/389
```

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_single_alerts_cmp.

ERROR: QA failed on SURI_TLPW2_autofp_alerts_cmp.

ERROR: QA failed on SURI_TLPW1_files_sha256.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .uptime | 147 | 353 | 240.14% |
| .tcp.rst | 113274 | 145060 | 128.06% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 1063 | 1496 | 140.73% |
| .tcp.overlap | 1126602 | 1222848 | 108.54% |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 993589 | 0.17 | 

Pipeline 13468

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_single_alerts_cmp.

ERROR: QA failed on SURI_TLPW2_autofp_alerts_cmp.

ERROR: QA failed on SURI_TLPW1_files_sha256.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .uptime | 147 | 347 | 236.05% |
| .tcp.rst | 113274 | 145252 | 128.23% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 1063 | 1500 | 141.11% |
| .tcp.overlap | 1126602 | 1222848 | 108.54% |
| .app_layer.error.snmp.parser | 1 | 0 | - |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 1008975 | 0.17 | 

Pipeline 13476

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_single_alerts_cmp.

ERROR: QA failed on SURI_TLPW2_autofp_alerts_cmp.

ERROR: QA failed on SURI_TLPW1_files_sha256.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .uptime | 147 | 348 | 236.73% |
| .tcp.rst | 113274 | 145037 | 128.04% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 1063 | 1502 | 141.3% |
| .tcp.overlap | 1126602 | 1222848 | 108.54% |
| .app_layer.error.snmp.parser | 1 | 0 | - |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 1018777 | 0.17 | 

Pipeline 13503

**Comment by victorjulien** [comment]
It looks like this also converts `http.response_body (sticky buffer)` / `http_server_body (content modifier)` are switched over to using files? These should continue to inspect the `HtpBody` object.

**Comment by victorjulien** [comment]
Could you rebase this and submit a new PR?

File: src/app-layer-htp-file.c [hunk 1]
```diff
@@ -66,6 +66,8 @@ int HTPFileOpen(HtpState *s, HtpTxUserData *tx, const uint8_t *filename, uint16_
         flags = FileFlowFlagsToFlags(tx->tx_data.file_flags, STREAM_TOSERVER);
     }
 
+    flags |= FILE_USE_DETECT;
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/app-layer-htp-file.c:70]: it makes file data sliding use 2 members of the file struct that http didn't update so far as it uses HtpBody

File: src/app-layer-htp.c [hunk 2]
```diff
@@ -1755,9 +1755,7 @@ static int HtpResponseBodyHandle(HtpState *hstate, HtpTxUserData *htud,
      * we check for tx->response_line in case of junk
      * interpreted as body before response line
      */
-    if (!(htud->tcflags & HTP_FILENAME_SET) &&
```
Comment (victorjulien) [inline src/app-layer-htp.c:1758]: these are unrelated changes?

File: src/app-layer-htp-file.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/app-layer-htp-file.c:70]: Should this flag not be on by default ?

I only see it used positively in `FilePruneFile` Am I right ?

File: src/app-layer-htp.c (hunk 2 above)
Comment (jlucovsky) [reply src/app-layer-htp.c:1769]: This was related to the change @catenacyber made: https://github.com/OISF/libhtp/commit/c629f266a47600f238fa10609c111e5b73831e2a

**Comment by jlucovsky** [comment]
> It looks like this also converts `http.response_body (sticky buffer)` / `http_server_body (content modifier)` are switched over to using files? These should continue to inspect the `HtpBody` object. Just `file.data` / `file_data` should use the `File` objects.

Will address.

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_single_alerts_cmp.

ERROR: QA failed on SURI_TLPW2_autofp_alerts_cmp.

ERROR: QA failed on SURI_TLPW1_files_sha256.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_single_stats_chk | | |
| .uptime | 994 | 2153 | 216.6% |
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 285 | 379 | 132.98% |
| | SURI_TLPW1_stats_chk | | |
| .uptime | 219 | 254 | 115.98% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 1241 | 1350 | 108.78% |
| .tcp.overlap | 1124925 | 1221167 | 108.56% |
| .flow.memuse | 592373632 | 521636440 | 88.06% | 

Pipeline 14957

File: src/app-layer-htp.c (hunk 2 above)
Comment (catenacyber) [reply src/app-layer-htp.c:1769]: This PR showed up a bug in libhtp which got fixed

The http-evader tests are based on files (check for eicar) with rules

As this PR changes file detection, the file needs to be opened (instead of just getting the response body)

This may be a separate commit, and maybe a S-V test to test for file info event instead of one with rules...

Comment (victorjulien) [reply src/app-layer-htp.c:1769]: Ok thanks. Yes lets get these in through a separate commit (and PR ?)

Comment (catenacyber) [reply src/app-layer-htp.c:1769]: @jlucovsky should I do it or do you ? (the separate PR/commit for these changes)

Comment (jlucovsky) [reply src/app-layer-htp.c:1769]: @catenacyber Yes -- thanks

Comment (catenacyber) [reply src/app-layer-htp.c:1769]: https://github.com/OISF/suricata/pull/9171

File: src/detect-file-data.c [hunk 3]
```diff
@@ -485,40 +229,121 @@ static InspectionBuffer *FiledataGetDataCallback(DetectEngineThreadCtx *det_ctx,
...
+
+        /* get the inspect buffer
+         *
+         * make sure that we have at least the configured inspect_win size.
+         * If we have more, take at least 1/4 of the inspect win size before
+         * the new data.
+         */
+        SCLogDebug("file_size %ld cur_file->content_inspected %ld cur_file->inspect_min_size %d",
+                file_size, cur_file->content_inspected, cur_file->inspect_min_size);
+        if (cur_file->content_inspected > cur_file->inspect_min_size) {
```
Comment (victorjulien) [inline src/detect-file-data.c:280]: this block is really hard to ready, can you clean it up, explain etc?

**Comment by victorjulien** [comment]
replaced by #9207
